src/herder/test/HerderTests.cpp

```diff
@@ -274,138 +287,145 @@ testTxSet(uint32 protocolVersion)
         nbTransactions * app->getLedgerManager().getLastTxFee() + minBalance0;
 
-    TxSetFramePtr txSet = std::make_shared<TxSetFrame>(Hash{});
+    TxSetFrameConstPtr txSet = std::make_shared<TxSetFrame const>(Hash{});
+    std::vector<TransactionFrameBasePtr>
+        txOriginal; // Keep the order of the generated transactions
 
     auto genTx = [&](int nbTxs) {
         std::string accountName = fmt::format("A{}", accounts.size());
         accounts.push_back(root.create(accountName.c_str(), amountPop));
         auto& account = accounts.back();
+        std::vector<TransactionFrameBasePtr> txs;
         for (int j = 0; j < nbTxs; j++)
         {
             // payment to self
-            txSet->add(account.tx({payment(account.getPublicKey(), 10000)}));
+            txs.push_back(account.tx({payment(account.getPublicKey(), 10000)}));
         }
+        txSet = TxSetUtils::addTxs(txSet, txs);
+        txOriginal.insert(txOriginal.end(), txs.begin(), txs.end());
     };
-
     for (size_t i = 0; i < nbAccounts; i++)
     {
         genTx(nbTransactions);
     }
-
-    txSet->previousLedgerHash() =
-        app->getLedgerManager().getLastClosedLedgerHeader().hash;
+    txSet = std::make_shared<TxSetFrame const>(
+        app->getLedgerManager().getLastClosedLedgerHeader().hash,
+        txSet->getTxsInHashOrder());
 
     SECTION("too many txs")
     {
-        while (txSet->mTransactions.size() <=
-               cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE)
+        while (txSet->sizeTx() <= cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE)
         {
             genTx(1);
         }
-        txSet->previousLedgerHash() =
-            app->getLedgerManager().getLastClosedLedgerHeader().hash;
-        txSet->sortForHash();
+        txSet = std::make_shared<TxSetFrame const>(
+            app->getLedgerManager().getLastClosedLedgerHeader().hash,
+            txSet->getTxsInHashOrder());
         REQUIRE(!txSet->checkValid(*app, 0, 0));
     }
     SECTION("order check")
     {
-        txSet->sortForHash();
-
         SECTION("success")
         {
             REQUIRE(txSet->checkValid(*app, 0, 0));
 
-            txSet->trimInvalid(*app, 0, 0);
+            trimInvalid(*app, txSet, 0, 0);
             REQUIRE(txSet->checkValid(*app, 0, 0));
-            REQUIRE(txSet->mTransactions.size() == (2 * nbTransactions));
+            REQUIRE(txSet->sizeTx() == (2 * nbTransactions));
         }
         SECTION("out of order")
         {
-            std::swap(txSet->mTransactions[0], txSet->mTransactions[1]);
+            TransactionSet xdrSet;
+            txSet->toXDR(xdrSet);
+            // rearrange it out of order
+            std::swap(xdrSet.txs[0], xdrSet.txs[1]);
+            txSet =
+                std::make_shared<TxSetFrame const>(app->getNetworkID(), xdrSet);
             REQUIRE(!txSet->checkValid(*app, 0, 0));
 
-            txSet->trimInvalid(*app, 0, 0);
+            // reconstruct it from TxSetFrame::Transactions, which automatically
```

Comment:
the goal here would be to test whatever code we're using in production when we perform `trimInvalid` --> another reason to have a `trimInvalid` equivalent

Reply:
Please see my earlier [comment](https://github.com/stellar/stellar-core/pull/3395#discussion_r862284708) on the `trimInvalid`.  `trimInvalid` does not sort-in-hash any more (the txSet was supposed to be sorted on construction), thus the scenario described here doesn't match the actual production. The closest thing to it is here where we take the out-of-order vector and reconstruct the `TxSet` from it to get a valid one.